### PR TITLE
Fix TFLite ArgMax/ArgMin NaN handling to match TensorFlow semantics

### LINKED ISSUE #118532: TFLite ArgMax returns different index from Keras when input contains NaN
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

3.11

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

TensorFlow Lite produces a different result from Keras for tf.math.argmax when the input contains NaN.

The second row of the input is [NaN, 7.0, NaN]. Keras returns index 1, corresponding to the finite value 7.0, while the converted TFLite model returns index 0.

In the reproducer below, Keras returns [0, 1], while TFLite returns [0, 0].

Expected behavior?

The converted TFLite model should match Keras behavior for tf.math.argmax on inputs containing NaN.

### Standalone code to reproduce the issue

```shell
import os
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

import warnings
warnings.filterwarnings("ignore")

import numpy as np
import tensorflow as tf


@tf.function
def fn(x):
    return tf.math.argmax(x, axis=1, output_type=tf.int32)


x = np.array(
    [[5.0, np.nan, 3.0],
     [np.nan, 7.0, np.nan]],
    dtype=np.float32,
)

eager = fn(tf.constant(x)).numpy()

cf = fn.get_concrete_function(tf.TensorSpec(x.shape, tf.float32))
tflite_model = tf.lite.TFLiteConverter.from_concrete_functions([cf]).convert()

interpreter = tf.lite.Interpreter(model_content=tflite_model)
interpreter.allocate_tensors()
interpreter.set_tensor(interpreter.get_input_details()[0]["index"], x)
interpreter.invoke()

tflite_out = interpreter.get_tensor(interpreter.get_output_details()[0]["index"])

print(f"input :\n{x}")
print(f"eager : {eager}")
print(f"tflite: {tflite_out}")

if eager[1] == 1 and tflite_out[1] != 1:
    print("BUG REPRODUCED: TFLite ArgMax returns different index from Keras")
else:
    print("not reproduced")
```

### Relevant log output

```shell
input :
[[ 5. nan 3.]
[nan 7. nan]]
eager : [0 1]
tflite: [0 0]
BUG REPRODUCED: TFLite ArgMax returns different index from Keras
```

### PR DESCRIPTION
**Pull request by CodersAcademy006**
## Summary

Fixes https://github.com/tensorflow/tensorflow/issues/118532

TFLite ArgMax/ArgMin produced incorrect indices for tensors containing NaN values due to IEEE 754 comparison semantics. When the reduction accumulator was initialized with a NaN value, subsequent finite candidates could not replace it because floating-point comparisons against NaN always evaluate to `false`.

This change aligns TFLite reduction behavior with TensorFlow eager semantics by ensuring finite values correctly replace NaN accumulators during ArgMax/ArgMin traversal while preserving existing tie semantics and optimized execution paths.

## Changes

- **`reference/arg_min_max.h`**: NaN-aware comparator via SFINAE for floating-point types. Non-float types unchanged.
- **`optimized/optimized_ops.h`**: NaN-aware mask computation in NEON loop body (`vbslq` replaces `vmaxq`/`vminq` to keep value/index synchronized). Scalar leftover loops use NaN-aware comparison. Vectorized horizontal reductions preserved unchanged.
- **`arg_min_max_test.cc`**: Focused regression coverage for mixed finite/NaN, all-NaN, and last-axis optimized paths.

## Test plan

- [ ] `bazel test //tensorflow/lite/kernels:arg_min_max_test`
- [ ] Verify NaN inputs return correct indices matching `tf.math.argmax`/`tf.math.argmin`
- [ ] Verify no regression on non-NaN float, int8, uint8, int32, bool inputs
- [ ] Verify deterministic index 0 for all-NaN inputs

### CODE REVIEW

**Comment by CodersAcademy006** [comment]
Rebased on latest upstream master. Build still shows 1 test failure; investigating the root cause to ensure CI passes on next run. Will ping for review once green.

**Comment by CodersAcademy006** [comment]
Added detailed NaN-aware comparison analysis: the code changes in ArgMinVector/ArgMaxVector NEON and scalar paths implement correct NaN handling per TensorFlow eager semantics (finite > NaN for reduction, all-NaN → index 0). The 1 test failure appears unrelated to the diff logic. Unable to verify with full local bazel build on available hardware. Recommend maintainers re-trigger CI or manually verify the failing test details. Logic review: reference_ops and optimized_ops now consistently handle NaNs.

**Comment by CodersAcademy006** [comment]
Friendly nudge — this is rebased onto latest `master` and ready for CI. The `CI` workflow is sitting in `action_required` (awaiting maintainer approval to run for an external contributor). Could a maintainer approve the workflow run so the build/tests can execute? Happy to address anything that comes up. Thanks!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thanks for working on this! The implementation itself looks solid, clean, and correctly matches TensorFlow eager semantics for NaN handling. 
  
  Before we can import this fix into the internal Google repository, could you please add additional unit tests to cover the reference execution path (where reduction is performed along a non-last axis)? Currently, the provided tests only exercise the optimized vectorized path (`inner_size == 1`).
  
  Here are the suggested unit tests you can append to `tensorflow/lite/kernels/arg_min_max_test.cc`:
  
  ```cpp
  // Reference path (non-last-axis) tests exercise GetCompareFunction and reference ArgMinMax.
  TEST_P(ArgMinMaxOpTest, ArgMaxNaNReference) {
    ArgMaxOpModel model({1, 1, 3, 2}, TensorType_FLOAT32, 2, AxisType(),
                        ConstantAxis(), OutputType());
    float nan = std::numeric_limits<float>::quiet_NaN();
    model.PopulateTensor<float>(model.input(), {nan, nan, 7.0f, 8.0f, nan, nan});
    ASSERT_EQ(model.Invoke(), kTfLiteOk);
    ValidateOutput(model, {1, 1});
  }
  
  TEST_P(ArgMinMaxOpTest, ArgMinNaNReference) {
    ArgMinOpModel model({1, 1, 3, 2}, TensorType_FLOAT32, 2, AxisType(),
                        ConstantAxis(), OutputType());
    float nan = std::numeric_limits<float>::quiet_NaN();
    model.PopulateTensor<float>(model.input(), {nan, nan, 7.0f, 8.0f, nan, nan});
    ASSERT_EQ(model.Invoke(), kTfLiteOk);
    ValidateOutput(model, {1, 1});
  }
  
  TEST_P(ArgMinMaxOpTest, ArgMaxAllNaNReference) {
    ArgMaxOpModel model({1, 1, 3, 2}, TensorType_FLOAT32, 2, AxisType(),
                        ConstantAxis(), OutputType());
    float nan = std::numeric_limits<float>::quiet_NaN();
    model.PopulateTensor<float>(model.input(), {nan, nan, nan, nan, nan, nan});
    ASSERT_EQ(model.Invoke(), kTfLiteOk);
    ValidateOutput(model, {0, 0});
  }
  
  TEST_P(ArgMinMaxOpTest, ArgMinAllNaNReference) {
    ArgMinOpModel model({1, 1, 3, 2}, TensorType_FLOAT32, 2, AxisType(),
                        ConstantAxis(), OutputType());
    float nan = std::numeric_limits<float>::quiet_NaN();
    model.PopulateTensor<float>(model.input(), {nan, nan, nan, nan, nan, nan});
    ASSERT_EQ(model.Invoke(), kTfLiteOk);
    ValidateOutput(model, {0, 0});
  }
  
  ```
Once those are added to the PR, we'll get this imported and merged.

**Comment by CodersAcademy006** [comment]
@dmiltr3 Added the four reference-path (non-last-axis, inner_size > 1) NaN tests for ArgMax and ArgMin in commit 46eaeb7; ready for import and merge.

**Review by dmiltr3 — APPROVED** [review]

**Review by dmiltr3 — APPROVED** [review]

**Comment by CodersAcademy006** [comment]
Friendly ping — this is approved (thanks @dmiltr3) and every build job is green except `build-windows-x86`, which fails on an unrelated proto-link flake in `quantize_op_test` (undefined `descriptor_table_...summary.proto` symbol; 1709 tests pass, 1 fails to build) rather than on the ArgMax/ArgMin change. Could this be imported/merged when you have a moment? Happy to rebase if anything has drifted.

**Comment by nithyak0204** [comment]
Hi @CodersAcademy006, Closing this PR as the proposed changes have already been merged into the codebase. Thank you for your contribution!
